---
title: "Entrez_download_experiment"
author: "Torda"
date: '2023 11 25 '
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


I think some simplification can be done on the download script, as well as length filtering and linking taxonomy and nucleotide database multiple time (the inconsistency) can be done before downloading sequences

```{r}
source("RSetup.R")
package.setup(workingdir = "/Users/varga/OneDrive/Documents/GitHub/PhyloGenie/TestFolder/")
search_term <- "Diehliomyces"
api_key <- NULL
path_to_output_dir <- "/Users/varga/OneDrive/Documents/GitHub/PhyloGenie/TestFolder/"

```

use Emily's example: Diehliomyces.
I think the entrez_post script can be skipped

```{r}
 r_search <- entrez_search(db = "taxonomy", 
                            term = paste0(search_term, "[subtree]"),
                            retmax = 999, 
                            use_history = TRUE) # gets all IDs for genus
ID_n <- length(r_search$ids) # reducing repetitions by reusing this variable
ID <- r_search$ids[1]
fetch_id <- entrez_link(dbfrom = "taxonomy", 
                                db = "nuccore",
                                id = ID)
summary_list <- entrez_summary("nuccore", fetch_id$links$taxonomy_nuccore)
```

from the summary statistics we can get the sequence length:
$slen
as well as the NCBI accession number:
$oslt$value
```{r}
key_info <- lapply(summary_list, function(x) c(x$uid, x$oslt$value, x$slen))
## filter for sequence length
lower <- 400
upper <- 10000
key_info[[1]][3]
select_idx <- unlist(lapply(key_info, function(x) as.numeric(x[3]) > lower & as.numeric(x[3]) < upper))
access_ID <- unlist(lapply(key_info[select_idx], function(x) x[2]))
fetch_id2 <- unlist(lapply(key_info[select_idx], function(x) x[1]))
```

I can put these lines into a loop to repeat if I can aquire more IDs

```{r}
lower <- 400
upper <- 10000
access_ID <- vector()
fetch_id2 <- vector()
for(i in 1:10){
fetch_id <- entrez_link(dbfrom = "taxonomy", 
                                db = "nuccore",
                                id = ID)
summary_list <- entrez_summary("nuccore", fetch_id$links$taxonomy_nuccore)
key_info <- lapply(summary_list, function(x) c(x$uid, x$oslt$value, x$slen))
select_idx <- unlist(lapply(key_info, function(x) as.numeric(x[3]) > lower & as.numeric(x[3]) < upper))
access_ID_temp <- unlist(lapply(key_info[select_idx], function(x) x[2]))
fetch_id2_temp <- unlist(lapply(key_info[select_idx], function(x) x[1]))
access_ID <- unique(c(access_ID, access_ID_temp))
fetch_id2 <- unique(c(fetch_id2, fetch_id2_temp))
cat("\nloop", i, "\t", length(access_ID))
}
```

